### PR TITLE
Added first draft of quality concept, illumina 1.8 implementation and test.

### PR DESCRIPTION
**Pull request by mariehoffmann**

### CODE REVIEW

File: include/seqan3/alphabet/quality.hpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+template <typename alphabet_type>
+requires detail::internal_quality_concept<alphabet_type>
+constexpr alphabet_char_t<alphabet_type> to_char(alphabet_type const & c)
+{
+    return c.to_phred();
+}
+
+
```
Comment (h-2) [inline include/seqan3/alphabet/quality.hpp]: please remove double newline!

File: include/seqan3/alphabet/quality.hpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+{
+    using type = typename alphabet_type::phred_type;
+};
+
+template <typename alphabet_type>
+using underlying_phred_t = typename underlying_phred<alphabet_type>::type;
+
+template <typename alphabet_type>
+requires detail::internal_quality_concept<alphabet_type>
+constexpr alphabet_type from_char(alphabet_type & c, char const in)
```
Comment (h-2) [inline include/seqan3/alphabet/quality.hpp]: it should be from_phred and to_phred, right?

File: include/seqan3/alphabet/quality.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    // offsets for char and integer encoding
+    q::offset_phred;
+    q::offset_char;
+    requires std::is_same_v<typename q::phred_type, std::decay_t<decltype(q::offset_phred)>>;
+    requires std::is_same_v<typename q::char_type, std::decay_t<decltype(q::offset_char)>>;
+
+    // conversion functions from t -> c and vice versa
+    { quality.from_phred(typename q::integral_type{}) } -> q;
+    { quality.to_phred() } -> const typename q::phred_type;
+
```
Comment (h-2) [inline include/seqan3/alphabet/quality.hpp]: change the concept so that it requires the global functions, instead of the member functions

File: include/seqan3/alphabet/quality.hpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+};
+
+}
```
Comment (h-2) [inline include/seqan3/alphabet/quality.hpp]: add newline at end of file

File: include/seqan3/alphabet/quality/phred.hpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+{
+
+using phred_type = int8_t;
+
+    struct illumina18
+    {
+        using char_type = char;
+        using integral_type = uint8_t;
+
+
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: remove double newline

File: include/seqan3/alphabet/quality/phred.hpp [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        static constexpr phred_type offset_phred{0};
+
+        // implicit compatibility to inner_type
+        constexpr illumina18 & operator =(integral_type const c)
+        {
+            value = c;
+        }
+
+        // comparison ops
+        // bool MyClass::operator==(const MyClass &other) const {
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: remove comment

File: include/seqan3/alphabet/quality/phred.hpp [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        constexpr bool operator ==(const illumina18 &rhs)
+        {
+            return this->value == rhs.value;
+        }
+
+        constexpr bool operator !=(const illumina18 &rhs)
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: our syntax for parameter is `illumina const & rhs` :nail_care: :smile:

File: include/seqan3/alphabet/quality/phred.hpp [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+            return *this;
+        }
+
+        // explicit compatibility to integral
+        constexpr integral_type to_integral() const
+        {
+            return value;
+        }
+
+        constexpr illumina18 from_integral(uint8_t const c)
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: use integral_type as parameter type, not uint8_t

File: include/seqan3/alphabet/quality/phred.hpp [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+    constexpr phred_type to_phred(illumina18 illu)
+    {
+        return illu.to_phred();
+    }
+
+    constexpr illumina18 from_phred(illumina18 illu, int8_t const p)
+    {
+        return illu.from_phred(p);
+    }
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: you don't need these anymore, if the global/free functions are fixed

File: test/alphabet/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,4 @@
+cmake_minimum_required(VERSION 3.6)
+project(alphabet_test CXX)
+
+add_subdirectory(quality)
```
Comment (h-2) [inline test/alphabet/CMakeLists.txt]: add newline to end of file

File: test/alphabet/quality/phred_test.cpp [hunk 11]
```diff
@@ -0,0 +1,85 @@
+#include <gtest/gtest.h>
+
+#include "../../../include/seqan3/alphabet/quality.hpp"
+#include "../../../include/seqan3/alphabet/quality/phred.hpp"
+
+using namespace seqan3;
+
+TEST( illumina18_constructor, default_ctr)
```
Comment (h-2) [inline test/alphabet/quality/phred_test.cpp]: remove space after opening `(`

File: test/alphabet/quality/phred_test.cpp [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    illumina18 illu2 = illu.from_char('!');
+    EXPECT_EQ(0, illu2.value);
+}
+
+
+TEST(illumina18_from_integral, from_integral)
+{
+    illumina18 illu;
+    illu = 0;
+illumina18 illu2 = illu.from_integral(1);
```
Comment (h-2) [inline test/alphabet/quality/phred_test.cpp]: indention?

File: test/alphabet/quality/phred_test.cpp [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    illumina18 illu2 = illu.from_phred(5);
+    EXPECT_EQ(5, illu2.value);
+}
+
+TEST(illumina18_to_phred, from_phred)
+{
+    illumina18 illu;
+    illumina18 illu2 = illu.from_phred(5);
+    EXPECT_EQ(5, illu2.value);
+}
```
Comment (h-2) [inline test/alphabet/quality/phred_test.cpp]: newline at end of file

**Comment by h-2** [comment]
also there are now conflicts, likely because of the testing stuff :o

File: include/seqan3/alphabet/quality.hpp [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+template <typename alphabet_type>
+requires detail::internal_quality_concept<alphabet_type>
+constexpr alphabet_type from_phred(alphabet_type & c, char const in)
+{
+    return c.from_phred(in);
+}
+
+template <typename alphabet_type>
+requires detail::internal_quality_concept<alphabet_type>
```
Comment (h-2) [inline include/seqan3/alphabet/quality.hpp]: please always indent the requires clauses

File: include/seqan3/alphabet/quality.hpp [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    // requires fulfillment of alphabet concept
+    requires alphabet_concept<q>;
+
+    // conversion functions from t -> c and vice versa
+    { from_phred(quality, typename q::integral_type{}) } -> q;
+    { to_phred(quality) } -> const typename q::phred_type;
+    typename underlying_phred<q>::type;
+};
+
+}
```
Comment (h-2) [inline include/seqan3/alphabet/quality.hpp]: please end a file with an empty newline

File: include/seqan3/alphabet/quality/phred.hpp [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+// Author: Marie Hoffmann <marie.hoffmann AT fu-berlin.de>
+// ============================================================================
+
+#pragma once
+
+#include <tuple>
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: why is this included?

File: include/seqan3/alphabet/quality/phred.hpp [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
+#include "../alphabet.hpp"
+#include "../quality.hpp"
+
+// assume sequential mapping w.r.t. ascii alphabet, only (char_start, phred_start) machine-dependent
+namespace seqan3
+{
+
+
+
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: please remove double newlines

File: include/seqan3/alphabet/quality/phred.hpp [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+    integral_type value;
+    // phred score intervals as int [0 .. 41], as char ['!' .. 'J']
+    static constexpr char_type offset_char{'!'};
+    static constexpr phred_type offset_phred{0};
+
+    // implicit compatibility to inner_type
+    constexpr illumina18 & operator =(integral_type const c)
+    {
+        value = c;
+    }
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: add `return *this;`

File: include/seqan3/alphabet/quality/phred.hpp [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+
+    // comparison ops
+    constexpr bool operator ==(const illumina18 & rhs)
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: the comparison operators need a ` const` in the end, too :)

File: include/seqan3/alphabet/quality/phred.hpp [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+
+    constexpr phred_type to_phred() const
+    {
+        return value + offset_phred;
+    }
+
+    static constexpr integral_type value_size{42};
+};
+
+    static_assert(quality_concept<illumina18>);
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: don't indent here!

File: test/alphabet/quality/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,12 @@
...
+
+include_directories("../gaps")
+include_directories("../nucleotide")
+include_directories("../quality")
+
+add_executable(test_phred phred_test.cpp)
+
+#add_dependencies(align_config_test gtest)
+target_link_libraries(test_phred ${SEQAN3_LIBRARIES})
+add_test(NAME test_phred_test COMMAND test_phred)
```
Comment (h-2) [inline test/alphabet/quality/CMakeLists.txt]: please end files always in a newline

File: test/alphabet/quality/phred_test.cpp [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+// DAMAGE.
+//
+// ============================================================================
+// Author: Marie Hoffmann <marie.hoffmann AT fu-berlin.de>
+// ============================================================================
+
+#include <gtest/gtest.h>
+
+#include "../../../include/seqan3/alphabet/quality.hpp"
+#include "../../../include/seqan3/alphabet/quality/phred.hpp"
```
Comment (h-2) [inline test/alphabet/quality/phred_test.cpp]: I think the tests should use absolute paths for access to SeqAn, e.g. `#include <seqan3/alphabet/quality.hpp>`

File: test/alphabet/quality/phred_test.cpp [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+
+using namespace seqan3;
+
```
Comment (h-2) [inline test/alphabet/quality/phred_test.cpp]: please test the public API, not the internal values, e.g.
`EXPECT_EQ(to_phred(illu), '!')`

constexpr values and can be checked with static_assert, e.g
`static_assert(alphabet_size<illumina18> == 42)`

File: include/seqan3/alphabet/quality/phred.hpp [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+// ============================================================================
+// Author: Marie Hoffmann <marie.hoffmann AT fu-berlin.de>
+// ============================================================================
+
+#pragma once
+
+#include "../alphabet.hpp"
+#include "../quality.hpp"
+
+
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: double newline

File: test/alphabet/quality/phred_test.cpp [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+// Author: Marie Hoffmann <marie.hoffmann AT fu-berlin.de>
+// ============================================================================
+
+#include <gtest/gtest.h>
+
+#include <seqan3/alphabet/quality.hpp>
+#include <seqan3/alphabet/quality/phred.hpp>
+
+using namespace seqan3;
+
```
Comment (h-2) [inline test/alphabet/quality/phred_test.cpp]: please check things like construction, copy construction et cetera!

File: test/alphabet/quality/phred_test.cpp [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+TEST(illumina18_constructor, const_offset)
+{
+    EXPECT_EQ(illumina18::offset_char, '!');
+}
+
+TEST(illumina18_alphabet_size, const_value_size)
+{
+    EXPECT_EQ(illumina18::value_size, 42);
```
Comment (h-2) [inline test/alphabet/quality/phred_test.cpp]: please test the "global" interface, i.e. `alphabet_size_v<illumina18>` in addition to `illumina18::value_size`

File: test/alphabet/quality/phred_test.cpp [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+TEST(illumina18_implicit_assign, implicit_assign)
+{
+    illumina18 illu;
+    illu = 19;
+    // expect size unmodified
+    EXPECT_EQ(illumina18::value_size, 42);
+    // newly assigned member
+    EXPECT_EQ(illu.value, 19);
```
Comment (h-2) [inline test/alphabet/quality/phred_test.cpp]: don't compare with `.value` because that is implementation detail, instead compare with `.to_integral()` and/or `to_integral(illu)`.

**Comment by h-2** [comment]
sorry, there are still some things, but we can look at it tomorrow together, too!

File: include/seqan3/alphabet/quality.hpp [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+    using type = typename alphabet_type::phred_type;
+};
+
+//! internal phred type
+template <typename alphabet_type>
+using underlying_phred_t = typename underlying_phred<alphabet_type>::type;
+
+//! public setter function receiving char encoding of phred score
+template <typename alphabet_type>
+requires detail::internal_quality_concept<alphabet_type>
```
Comment (smehringer) [inline include/seqan3/alphabet/quality.hpp]: indent requires statement :)

File: include/seqan3/alphabet/quality/phred.hpp [hunk 29]
```diff
@@ -0,0 +1,155 @@
...
+    {
+        return value + offset_phred;
+    }
+
+    //! phred score range for Illumina 1.8 standard
+    static constexpr integral_type value_size{42};
+};
+
+//! assert when (internal) quality concept requirements are not met
+static_assert(quality_concept<illumina18>);
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: One last thing: please wrap this in a #ifndef NDEBUG block! see the other alphabets for more details!

File: include/seqan3/alphabet/quality/phred.hpp [hunk 30]
```diff
@@ -0,0 +1,155 @@
...
+    //! explicit compatibility to internal integral representation
+    constexpr integral_type to_integral() const
+    {
+        return value;
+    }
+
+    //! set internal value given zero-based integer c
+    constexpr illumina18 from_integral(integral_type const c)
+    {
+        value = c;
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: add an assert to check that the integral is smaller than the value size

File: include/seqan3/alphabet/quality/phred.hpp [hunk 31]
```diff
@@ -0,0 +1,155 @@
...
+        return *this;
+    }
+
+    //! set internal value given Illumina 1.8 integer code p
+    constexpr illumina18 from_phred(phred_type const p)
+    {
+        value = p - offset_phred;
```
Comment (h-2) [inline include/seqan3/alphabet/quality/phred.hpp]: add an assert that the phred value is in the allowed range!
